### PR TITLE
 [CMSIS-NN] Fixed the case with repeating operands in the QNN binary ops

### PR DESCRIPTION
**Pull request by ashutosh-arm**
This commit is to fix issues in CMSIS-NN passes
that surface when the same operand is repeated
in QNN binary ops.

cc @areusch

### CODE REVIEW

**Comment by ashutosh-arm** [comment]
cc: @Mousius @grant-arm @manupa-arm for code review.

File: python/tvm/relay/op/contrib/cmsisnn.py [hunk 1]
```diff
@@ -229,17 +229,18 @@ def qnn_max_pool2d_pattern():
     def check_qnn_max_pool2d(pattern):
         """Check if max pool2d is supported by CMSIS-NN."""
         output = pattern
-        input_var = _find_last(pattern)
+        input_op = None
```
Comment (Mousius) [inline python/tvm/relay/op/contrib/cmsisnn.py:232]: We can remove this as it's always set later.

File: tests/python/contrib/test_cmsisnn/test_binary_ops.py [hunk 2]
```diff
@@ -145,6 +145,58 @@ def test_op_int8(
...
+    # validate pattern matching
+    assert_partitioned_function(orig_mod, cmsisnn_mod)
+
+    # validate the output
+    in_min, in_max = get_range_for_dtype_str(dtype)
+    inputs = {
+        "input": np.random.randint(in_min, high=in_max, size=shape, dtype=dtype),
+    }
+    output_list = generate_ref_data(orig_mod["main"], inputs)
+    compile_and_run(
```
Comment (Mousius) [inline tests/python/contrib/test_cmsisnn/test_binary_ops.py:187]: Should we not test that the internal functions only have 1 parameter each?

File: tests/python/contrib/test_cmsisnn/test_scalar_to_tensor_constant.py [hunk 3]
```diff
@@ -256,6 +256,47 @@ def test_all_primary_operands_tensor_constants():
     assert tvm.ir.structural_equal(mod[global_var].body, new_mod[global_var].body)
 
 
+@tvm.testing.requires_cmsisnn
+def test_duplicate_constant_arguments():
+    """Tests the pass when repeating operands are arguments to the binary op"""
+    dtype = "int8"
+    shape = (1, 3, 3, 32)
+    operand0 = generate_variable("operand0", shape, dtype)
+    operand1 = generate_variable("operand0", shape, dtype)
```
Comment (Mousius) [inline tests/python/contrib/test_cmsisnn/test_scalar_to_tensor_constant.py:265]:
```suggestion
    operand0 = generate_variable("operand0", shape, dtype)
    operand1 = generate_variable("operand1", shape, dtype)
```

File: tests/python/contrib/test_cmsisnn/test_scalar_to_tensor_constant.py [hunk 4]
```diff
@@ -256,6 +256,47 @@ def test_all_primary_operands_tensor_constants():
...
+    main_var = relay.GlobalVar("main")
+
+    mod = tvm.IRModule()
+    mod[global_var] = extern_func
+    mod[main_var] = main_func
+
+    mod = relay.transform.InferType()(mod)
+    mod = ScalarToTensorConstants()(mod)
+    new_mod = relay.transform.InferType()(mod)
+    assert tvm.ir.structural_equal(mod[global_var].body, new_mod[global_var].body)
```
Comment (Mousius) [inline tests/python/contrib/test_cmsisnn/test_scalar_to_tensor_constant.py:297]: What is this checking? It appears to just check the body hasn't changed after `InferType`?

**Review by Mousius — CHANGES_REQUESTED** [review]
I have a few questions @ashutosh-arm, hopefully I've not missed the obvious 😸

File: python/tvm/relay/op/contrib/cmsisnn.py (hunk 1 above)
Comment (ashutosh-arm) [reply python/tvm/relay/op/contrib/cmsisnn.py:232]: ACK

File: tests/python/contrib/test_cmsisnn/test_scalar_to_tensor_constant.py (hunk 4 above)
Comment (ashutosh-arm) [reply tests/python/contrib/test_cmsisnn/test_scalar_to_tensor_constant.py:297]: From offline discussion it was decided to leave the check for structural equality in there.

File: tests/python/contrib/test_cmsisnn/test_binary_ops.py (hunk 2 above)
Comment (ashutosh-arm) [reply tests/python/contrib/test_cmsisnn/test_binary_ops.py:194]: I have added a check for this. Thanks @Mousius.

**Comment by ashutosh-arm** [comment]
Thanks @Mousius for the discussions and code review. Would you like to take a look at it again?

**Comment by Mousius** [comment]
Thanks @ashutosh-arm 😸

**Pull request merged**